### PR TITLE
Hide "Rescan Drone Factory Folder" if no Drone Factory folder is setup

### PR DESCRIPTION
**Pull request by JoshSchreuder**
#### Database Migration
NO

#### Description
Hides the "Rescan Drone Factory Folder" on the Wanted page if the user has no Drone Factory folder configured

#### Todos
- [x] Tests
- [n/a] Documentation

#### Issues Fixed or Closed by this PR

* #2009 (up-for-grabs)

### CODE REVIEW

File: src/NzbDrone.Automation.Test/NzbDrone.Automation.Test.csproj [hunk 1]
```diff
@@ -58,16 +58,17 @@
...
     <Reference Include="Microsoft.CSharp" />
-    <Reference Include="WebDriver, Version=3.2.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\packages\Selenium.WebDriver.3.2.0\lib\net40\WebDriver.dll</HintPath>
+    <Reference Include="WebDriver, Version=3.8.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\Selenium.WebDriver.3.8.0\lib\net40\WebDriver.dll</HintPath>
     </Reference>
-    <Reference Include="WebDriver.Support, Version=3.2.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\packages\Selenium.Support.3.2.0\lib\net40\WebDriver.Support.dll</HintPath>
+    <Reference Include="WebDriver.Support, Version=3.8.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\Selenium.Support.3.8.0\lib\net40\WebDriver.Support.dll</HintPath>
```
Comment (JoshSchreuder) [inline src/NzbDrone.Automation.Test/NzbDrone.Automation.Test.csproj]: Apologies on this - I was unable to get WebDriver running locally with FF 57 unless I upgraded this. Happy to revert this change if you want.

File: src/UI/Wanted/Missing/MissingLayout.js [hunk 2]
```diff
@@ -77,8 +79,16 @@ module.exports = Marionette.Layout.extend({
...
     onShow : function() {
+        var self = this;
+
+        this.downloadClientSettings = new DownloadClientSettingsModel();
         this.missing.show(new LoadingView());
-        this._showToolbar();
+
+        Backbone.$.when(this.downloadClientSettings.fetch()).done(function() {
+            self.droneFolder = self.downloadClientSettings.attributes.downloadedEpisodesFolder;
+            self._showToolbar();
```
Comment (JoshSchreuder) [inline src/UI/Wanted/Missing/MissingLayout.js]: This does add a slight delay while the settings are fetched before rendering the toolbar - feedback appreciated on if this is the best way of going about it as I wasn't too sure.

File: src/NzbDrone.Automation.Test/WantedPageTest.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+        [Test]
+        public void show_drone_update_if_drone_folder_set()
+        {
+            page.SettingNavIcon.Click();
+            page.WaitForNoSpinner();
+
+            page.FindByClass("x-download-client-tab").Click();
+            page.FindByClass("x-advanced-settings").Click();
+            page.Find(By.Name("downloadedEpisodesFolder")).SendKeys("C:\\");
```
Comment (JoshSchreuder) [inline src/NzbDrone.Automation.Test/WantedPageTest.cs]: Is there a drone folder value I can use that lets this test pass cross-platform?

File: src/NzbDrone.Automation.Test/PageModel/PageBase.cs [hunk 4]
```diff
@@ -27,6 +27,11 @@ public IWebElement Find(By by, int timeout = 5)
             return wait.Until(d => d.FindElement(by));
         }
 
+        public bool Displayed(By by)
+        {
+            return _driver.FindElement(by).Displayed;
```
Comment (JoshSchreuder) [inline src/NzbDrone.Automation.Test/PageModel/PageBase.cs]: Added this so I can assert the actual exception rather than the wait timeout exception from the above wrapper methods.

File: src/NzbDrone.Automation.Test/WantedPageTest.cs (hunk 3 above)
Comment (Taloth) [reply src/NzbDrone.Automation.Test/WantedPageTest.cs]: There's an `.AsOsAgnostic()` extension method, but I don't think that'll work coz the validator checks if the folder exists. Better just create a temporary folder somewhere. The NzbDroneRunner instance has an AppData property that contains a temp folder just for that test. So create a subfolder there and use it as fake drone factory folder.
But don't spend too much time on those automation tests. Atm they only test whether the UI shows up and the main pages are browsable, little more than a smoketest.

File: src/NzbDrone.Automation.Test/NzbDrone.Automation.Test.csproj (hunk 1 above)
Comment (Taloth) [reply src/NzbDrone.Automation.Test/NzbDrone.Automation.Test.csproj]: Build agent seemed to be happy with 3.8 so 👍

File: src/UI/Wanted/Missing/MissingLayout.js [hunk 5]
```diff
@@ -144,6 +148,16 @@ module.exports = Marionette.Layout.extend({
                 }
             ]
         };
+
+        if (this.droneFolder) {
+            leftSideButtons.items.splice(4, 0, {
```
Comment (Taloth) [inline src/UI/Wanted/Missing/MissingLayout.js]: I don't like this magically splicing in the button. Options:
- Append Manual Import as well
- Keep as array, and remove the Drone Factory by findIndex.
- Add 'hidden' property and extend ToolbarLayout to skip it if hidden is or evaluates to true.

File: src/UI/Wanted/Missing/MissingLayout.js (hunk 2 above)
Comment (Taloth) [reply src/UI/Wanted/Missing/MissingLayout.js]: @markus101 This one's for you, I don't know what's the best way to deal with this.

Comment (markus101) [reply src/UI/Wanted/Missing/MissingLayout.js]: A few ideas:

- Fetch download client settings when the app loads, much like UI settings and ensure it stays in sync when changes are made
- Fetch it when loading Wanted: Missing or settings and update it if the settings are changes. If it never changes it's cached, if it does, we can re-render and remove the button.
- Fetch during initialize then re-rendering after the request completes (if the display of the button would change)

There could be some re-rendering of the buttons if the visibility changes after the initial showing, but if we moved that button the the last position it'd just be added/removed (minus some minor styling changes)

Comment (Taloth) [reply src/UI/Wanted/Missing/MissingLayout.js]: The caching approach is definitely the best, although I wonder if it's worth it considering v3 gets rid of the button entirely.

Two options that don't involve caching stuff:
1. Initial render with button disabled (not removed). Fetch and re-render (enabling the button) if needed.
2. Initial render with button removed. Fetch and re-render (adding the button) if needed.

1 has the least flicker, 2 is useful for most users since by now they shouldn't have the DF anymore.
